src/plugins/mer/meremulatormodeoptionswidget.cpp

```diff
@@ -0,0 +1,181 @@
+            this, &MerEmulatorModeOptionsWidget::deviceModelScreenChanged);
+    connect(m_ui->screenHeightSpinBox, &QAbstractSpinBox::editingFinished,
+            this, &MerEmulatorModeOptionsWidget::deviceModelScreenChanged);
+    connect(m_ui->screenWidthSpinBox, &QAbstractSpinBox::editingFinished,
+            this, &MerEmulatorModeOptionsWidget::deviceModelScreenChanged);
+    connect(m_ui->dconfWidget, &MerEmulatorDeviceDconfOptionsWidget::optionsChanged,
+            this, &MerEmulatorModeOptionsWidget::deviceModelDconfChanged);
+    connect(m_ui->deviceModelComboBox, qOverload<int>(&QComboBox::currentIndexChanged),
+            this, &MerEmulatorModeOptionsWidget::updateGui);
+    connect(m_ui->nameLineEdit, &QLineEdit::textEdited,
+            this, &MerEmulatorModeOptionsWidget::deviceModelNameChanged);
+
+    updateGui();
+}
+
+MerEmulatorModeOptionsWidget::~MerEmulatorModeOptionsWidget()
+{
+    MerSettings::deviceModelsReset();
+
+    delete m_ui;
+}
+
+void MerEmulatorModeOptionsWidget::store()
+{
+    MerSettings::deviceModelsStore();
+}
+
+QString MerEmulatorModeOptionsWidget::searchKeywords() const
+{
+    const QStringList keywordsList {
+        m_ui->deviceModelLabel->text(),
+        m_ui->screenResolutionLabel->text(),
+        m_ui->screenSizeLabel->text(),
+        m_ui->dconfLabel->text(),
+        m_ui->addProfileButton->toolTip(),
+        m_ui->removeProfileButton->toolTip()
+    };
+
+    QString keywords = keywordsList.join(' ');
+    keywords.remove(QLatin1Char('&'));
+    return keywords;
+}
+
+void MerEmulatorModeOptionsWidget::deviceModelNameChanged(const QString &newName)
+{
+    QPalette p = m_ui->nameLineEdit->palette();
+    QColor color = Qt::black;
+    QString toolTip = tr("Rename device model");
+
+    const QString name = m_ui->deviceModelComboBox->currentDeviceModel();
+    if (MerSettings::renameDeviceModel(name, newName)) {
+        m_ui->nameLineEdit->setFocus();
+    } else {
+        if (name != newName) {
+            color = Qt::red;
+            toolTip = tr("Device model name is not unique");
+        }
+    }
+
+    p.setColor(QPalette::Text, color);
+    m_ui->nameLineEdit->setPalette(p);
+    m_ui->nameLineEdit->setToolTip(toolTip);
+}
+
+void MerEmulatorModeOptionsWidget::deviceModelDconfChanged(const QString &value)
+{
+    const QString name = m_ui->deviceModelComboBox->currentDeviceModel();
+    const MerEmulatorDeviceModel model = MerSettings::deviceModels().value(name);
+    QTC_ASSERT(!model.isNull(), return);
+
+    if (model.dconf() != value) {
+        MerSettings::updateDeviceModel(MerEmulatorDeviceModel(name,
+                                                              model.displayResolution(),
+                                                              model.displaySize(),
+                                                              value,
+                                                              model.isSdkProvided()));
```

Comment:
Only now I noticed that the options page actually modifies the "real" data immediately and only saving to config files is delayed. It should work on a copy of the data instead. Is there any example of such an approach in upstream qtc code? I am only aware of code that uses temporary data.

Reply:
I did it on purpose. The reason is: user may change device model, then start emulator without leaving Options dialog widget to test it. All changes will reset if user cancel Options dialog in
`MerEmulatorModeOptionsWidget::~MerEmulatorModeOptionsWidget()`.
This approach, I believe, produce less code than fiddling with copies.

Reply:
I am sorry but unless you point me to some upstream code that does it similarly, I must insist on working on temporary data and only change the actual data when Apply/OK is hit. It requires more code, but this is the approach used in existing code that I am aware off. The use case you described would still be possible through the use of the Apply button :)

Reply:
Done. Work with copy now.